plugin/trino-exchange/src/main/java/io/trino/plugin/exchange/FileSystemExchangeSource.java

```diff
@@ -72,44 +89,63 @@ public synchronized Slice read()
         if (isFinished()) {
             return null;
         }
+        Slice sizeSlice = Slices.allocate(Integer.BYTES);
+        readFromExchangeStorage(sizeSlice);
+        int size = sizeSlice.getInt(0);
+        Slice dataSlice = Slices.allocate(size);
+        readFromExchangeStorage(dataSlice);
+        return dataSlice;
+    }
 
-        if (sliceInput != null && !sliceInput.isReadable()) {
-            sliceInput.close();
-            sliceInput = null;
+    private void readFromExchangeStorage(Slice slice)
+    {
+        readIfNeeded();
+        int position = 0;
+        while (position < slice.length()) {
+            int readableBytes = min(sliceInput.available(), slice.length() - position);
+            slice.setBytes(position, sliceInput.readSlice(readableBytes));
+            position += readableBytes;
+
+            readIfNeeded();
         }
+    }
 
-        if (sliceInput == null) {
-            if (files.hasNext()) {
-                // TODO: implement parallel read
-                URI file = files.next();
-                Optional<SecretKey> secretKey = secretKeys.next();
+    private void readIfNeeded()
+    {
+        if (sliceInput == null || !sliceInput.isReadable()) {
+            // swap dataBuffer and readBuffer
+            Slice tempBuffer = dataBuffer;
+            dataBuffer = readBuffer;
+            readBuffer = tempBuffer;
```

Comment:
It feels like having just two buffers here, whihc we swap does not help a lot if we still need to do data copying in `read()` where we are building Slices.
IIUC this is what you want to address with prod-cons approach, right?

Reply:
You are right, that's partially why I want to switch to the producer-consumer approach. Another advantage is we can do parallel reading too.